### PR TITLE
Composition postgres 142

### LINKED ISSUE #142: Composition instead of inheritance for PostgresJobStore
We're dealing with fine software design issues, it's really no big deal if we don't go this way.

I think we could make a last (haha, last...) modification on how our job store handles being plugged with a real `Postgres` database versus being in memory for tests and such . Today, it's done via inheritance, with a `BaseJobStore`, a `PostgresJobStore` and an `ImMemoryJobStore`. I've finally realized that it would be much cleaner in terms of separation of concerns if the whole idea of store was uncorrelated from the idea of DB connection. We could have a `JobStore`, and a `PostgresConnector` and an `InMemoryConnector`, and the `JobStore` would receive an `...Connector` instance at creation time (same way as the App currently receives a Job Store instance).

That would mean the App would be given a connector instead of a `JobStore`. If we change the name, that would be backwards incompatible, but we could still keep the name `PostgresJobStore` as an alias of `PostgresConnector` for at least a few releases.

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #142 

Includes breaking changes (but I kept a compat layer and added a deprecation warning)

### Successful PR Checklist:
- [x] Tests
- [x] Documentation (optionally: [run spell checking](https://github.com/peopledoc/procrastinate/blob/master/CONTRIBUTING.rst#build-the-documentation))
- [x] Had a good time contributing? (if not, feel free to give some feedback) < **Not as bad as expected :D**

### CODE REVIEW

File: procrastinate/aiopg_connector.py [hunk 1]
```diff
@@ -131,3 +131,12 @@ def interrupt_wait(self):
...
 
+
+class PostgresJobStore(PostgresConnector):
+    def __init__(self, *args, **kwargs):
+        message = (
+            "Use procrastinate.PostgresConnector(...) "
+            "instead of procrastinate.PostgresJobStore(...), with the same arguments"
+        )
+        logger.warn(f"Deprecation Warning: {message}")
+        warnings.warn(DeprecationWarning(message))
```
Comment (ewjoachim) [inline procrastinate/aiopg_connector.py]: super()

File: procrastinate/store.py [hunk 2]
```diff
@@ -13,37 +11,22 @@ def get_channel_for_queues(queues: Optional[Iterable[str]] = None) -> Iterable[s
         return ["procrastinate_queue#" + queue for queue in queues]
 
 
-class BaseJobStore:
+class JobStore:
     json_dumps: Optional[Callable] = None
     json_loads: Optional[Callable] = None
```
Comment (elemoine) [inline procrastinate/store.py]: To be removed.

File: procrastinate/app.py [hunk 3]
```diff
@@ -209,11 +214,15 @@ def perform_import_paths(self):
 
     @property
     def migrator(self) -> migration.Migrator:
-        return migration.Migrator(job_store=self.job_store)
+        return migration.Migrator(connector=self.connector)
+
+    @property
+    def job_store(self) -> store.JobStore:
+        return store.JobStore(connector=self.connector)
```
Comment (ewjoachim) [inline procrastinate/app.py]: instantiate in constructor

File: tests/integration/test_pg_connector.py [hunk 4]
```diff
@@ -8,42 +8,45 @@
 pytestmark = pytest.mark.asyncio
 
 
-async def test_get_connection(connection):
-    dsn = connection.dsn
-    async with await aiopg_connector.get_connection(dsn=dsn) as new_connection:
+@pytest.fixture
```
Comment (ewjoachim) [inline tests/integration/test_pg_connector.py]: rename file as test_aiopg_connector

**Comment by ewjoachim** [comment]
@elemoine @k4nar Shall I fixup my fixups ?

**Comment by k4nar** [comment]
You can squash them if you want.

**Comment by ewjoachim** [comment]
LGTY ?

File: docs/howto/testing.rst [hunk 5]
```diff
@@ -1,20 +1,20 @@
 Test your code that uses Procrastinate
 --------------------------------------
 
-Procrastinate defines an `InMemoryJobStore` that will speed up your tests,
-remove dependency to PostgreSQL and allow you to have tasks run in a
+Procrastinate defines an `InMemoryConnector` that will speed up your tests,
+remove dependency to PostgreSQL and allows you to have tasks run in a
```
Comment (elemoine) [inline docs/howto/testing.rst]: s/allows/allow, because of `will` earlier in the sentence.

File: procrastinate/aiopg_connector.py [hunk 6]
```diff
@@ -125,60 +44,100 @@ def __init__(
...
+    def interrupt_wait(self):
+        if not self._connection:
+            return
+        asyncio.get_event_loop().call_soon_threadsafe(
+            self._connection.notifies.put_nowait, "s"
         )
 
-    def stop(self):
-        if self._connection:
-            interrupt_wait(connection=self._connection)
```
Comment (elemoine) [inline procrastinate/aiopg_connector.py]: Did you purposefully remove `stop` here?

File: procrastinate/app.py [hunk 7]
```diff
@@ -51,13 +57,29 @@ def __init__(
             A :py:func:`App.task` that has a custom "name" parameter, that is not
             imported and whose module path is not in this list will
             fail to run.
+        job_store:
+            **Deprecated**: Old name of ``connector``
```
Comment (elemoine) [inline procrastinate/app.py]: Final dot for consistency with the other docstrings.

File: procrastinate/aiopg_connector.py (hunk 6 above)
Comment (ewjoachim) [reply procrastinate/aiopg_connector.py]: It has moved to the store: https://github.com/peopledoc/procrastinate/pull/147/files/9683ef2955cf5ba9f0128c4220f200c347007167#diff-32d29d678d9b442460d4b55326ed2104R23

File: procrastinate/app.py [hunk 8]
```diff
@@ -51,13 +57,29 @@ def __init__(
...
+        if job_store:
+            message = (
+                "Use App(connector=procrastinate.PostgresConnector(...)) "
+                "instead of App(job_store=procrastinate.PostgresJobStore())"
+            )
+            logger.warn(f"Deprecation Warning: {message}")
+            warnings.warn(DeprecationWarning(message))
+            connector = job_store
+        if not connector:
+            raise TypeError("App() missing 1 required argument: 'connector'")
```
Comment (elemoine) [inline procrastinate/app.py]: Do we want to raise an Exception when connector and job_store are both set in the constructor arguments?

File: tests/integration/test_pg_connector.py (hunk 4 above)
Comment (elemoine) [reply tests/integration/test_pg_connector.py]: do you plan to do it with a separate PR?

File: procrastinate/app.py (hunk 8 above)
Comment (ewjoachim) [reply procrastinate/app.py]: Do you think people could end up being confused to the point they submit those 2 parameters and are not happy with the result ?

File: procrastinate/aiopg_connector.py [hunk 9]
```diff
@@ -125,60 +44,100 @@ def __init__(
...
+            )
+        except asyncio.TimeoutError:
+            pass
 
-    async def wait_for_jobs(self):
-        return await wait_for_jobs(
-            connection=await self.get_connection(), socket_timeout=self.socket_timeout
+    # This can be called from a signal handler, better not do async stuff
+    def interrupt_wait(self):
+        if not self._connection:
```
Comment (elemoine) [inline procrastinate/aiopg_connector.py]: I think we need to have `if not self._connection or self._connection.closed` here.

File: procrastinate/app.py (hunk 8 above)
Comment (elemoine) [reply procrastinate/app.py]: Not sure. Up to you really.

File: tests/integration/test_pg_connector.py (hunk 4 above)
Comment (ewjoachim) [reply tests/integration/test_pg_connector.py]: Damn, you're right

File: procrastinate/app.py (hunk 8 above)
Comment (ewjoachim) [reply procrastinate/app.py]: I vote for keeping the code simple and avoiding having one more test to write :)

File: procrastinate/aiopg_connector.py (hunk 6 above)
Comment (elemoine) [reply procrastinate/aiopg_connector.py]: In that case there shouldn't be a `stop` method in the `InMemoryConnector` either, no?

**Pull request merged**